src/core/namespace.c

```diff
@@ -3291,6 +3307,467 @@ bool ns_type_supported(NamespaceType type) {
         return access(ns_proc, F_OK) == 0;
 }
 
+static int unpeel_get_fd(const char *mount_path, int *ret_fd) {
+        _cleanup_close_pair_ int pipe_fds[2] = EBADF_PAIR;
+        _cleanup_close_ int fs_fd = -EBADF;
+        pid_t pid;
+        int r;
+
+        assert(mount_path);
+        assert(ret_fd);
+
+        r = socketpair(AF_UNIX, SOCK_DGRAM|SOCK_CLOEXEC, 0, pipe_fds);
+        if (r < 0)
+                return log_debug_errno(errno, "Failed to create socket pair: %m");
+
+        /* Clone mount namespace here to unpeel without affecting live process */
+        r = safe_fork("(sd-ns-unpeel)", FORK_DEATHSIG_SIGTERM|FORK_LOG|FORK_WAIT|FORK_NEW_MOUNTNS|FORK_MOUNTNS_SLAVE, &pid);
+        if (r < 0)
+                return r;
+        if (r == 0) {
+                _cleanup_close_ int dir_fd = -EBADF;
+
+                pipe_fds[0] = safe_close(pipe_fds[0]);
+
+                /* Opportunistically unmount any overlay at this path */
+                r = path_is_fs_type(mount_path, OVERLAYFS_SUPER_MAGIC);
```

Comment:
as mentioned this will fall apart for anyone who uses overlayfs as the root fs for other reasons. i.e. anyone doing systemd.volatile= or using composefs or whatnot. this must be a more precise check, so that we recognized our own and stop once we hit somebody else's overlayfs

Reply:
here's an idea to address this quite reliably: set a recognizable "source" field for the mount, that makes this recognizable. i.e. set it to "systemd-sysext-unit" and "systemd-confext-unit" or so, and check that here. statmount() should make this very efficient and would be a very realiable way to recognize our own mounts here

Reply:
I've implemented that with the existing path_get_mount_info() helper, as statmount unfortunately is a bit too new (6.8, I'm using the 6.6 LTS). Once we move past 6.6 we can update the helper to use statmount and this will benefit too. But otherwise using the source field seems to work